src/useCameraShake.tsx

```diff
@@ -0,0 +1,125 @@
+  decay: true,
+  decayRate: 0.75,
+  maxYaw: 0.1,
+  maxPitch: 0.1,
+  maxRoll: 0.1,
+  yawFrequency: 10,
+  pitchFrequency: 10,
+  rollFrequency: 10,
+  yawNoiseSeed: 10,
+  pitchNoiseSeed: 20,
+  rollNoiseSeed: 30,
+}
+
+export interface ShakeController {
+  setTrauma: (val: number) => void
+  getTrauma: () => number
+  addTrauma: (val: number) => void
+  clearTrauma: () => void
+  updateConfig: (cfg: ShakeConfigPartial) => void
+}
+
+export function useCameraShake(
+  controlledCam: React.MutableRefObject<Camera | undefined>,
+  shakeConfig?: ShakeConfigPartial,
+  initialTrauma?: number
+) {
+  const { setDefaultCamera } = useThree()
+  const [config, setConfig] = React.useState<ShakeConfig>({ ...defaultConfig, ...shakeConfig })
+  const shakyCam = React.useRef<Camera>(new THREE.PerspectiveCamera())
+  const trauma = React.useRef<number>(initialTrauma ? initialTrauma : 0) // range [0-1]
+
+  const yawNoise = React.useRef<Noise2D>(makeNoise2D(config.yawNoiseSeed))
+  const pitchNoise = React.useRef<Noise2D>(makeNoise2D(config.pitchNoiseSeed))
+  const rollNoise = React.useRef<Noise2D>(makeNoise2D(config.rollNoiseSeed))
+
+  const constrainTrauma = () => {
+    if (trauma.current < 0 || trauma.current > 1) {
+      trauma.current = trauma.current < 0 ? 0 : 1
+    }
+  }
+
+  const setTrauma = (val: number) => {
+    trauma.current = val
+    constrainTrauma()
+  }
+
+  const getTrauma = () => {
+    return trauma.current
+  }
+
+  const addTrauma = (val: number) => {
+    trauma.current += val
+    constrainTrauma()
+  }
+
+  const clearTrauma = () => {
+    trauma.current = 0
+  }
+
+  const updateConfig = (cfg: ShakeConfigPartial) => {
+    setConfig({ ...config, ...cfg })
+  }
+
+  React.useEffect(() => {
+    yawNoise.current = makeNoise2D(config.yawNoiseSeed)
+  }, [config.yawNoiseSeed])
+
+  React.useEffect(() => {
+    pitchNoise.current = makeNoise2D(config.pitchNoiseSeed)
+  }, [config.pitchNoiseSeed])
+
+  React.useEffect(() => {
+    rollNoise.current = makeNoise2D(config.rollNoiseSeed)
+  }, [config.rollNoiseSeed])
+
+  useFrame(({ clock }, delta) => {
+    if (!controlledCam.current) return // ref checks
+
+    if (trauma.current > 0) {
+      shakyCam.current.copy(controlledCam.current as never) // I dont understand why this expects never.
+      const shake = Math.pow(trauma.current, 2)
+
+      const yaw = config.maxYaw * shake * yawNoise.current(clock.elapsedTime * config.yawFrequency, 1)
+      const pitch = config.maxPitch * shake * pitchNoise.current(clock.elapsedTime * config.pitchFrequency, 1)
+      const roll = config.maxRoll * shake * rollNoise.current(clock.elapsedTime * config.rollFrequency, 1)
+
+      shakyCam.current.rotation.x += pitch
+      shakyCam.current.rotation.y += yaw
+      shakyCam.current.rotation.z += roll
+
+      if (config.decay) {
+        trauma.current -= config.decayRate * delta
+        constrainTrauma()
+      }
+
+      setDefaultCamera(shakyCam.current)
```

Comment:
I think paul has mentioned before that `setDefaultCamera` shouldn't be called in useFrame.

Reply:
Hmm ok, I wasn't aware of that. It seemed like it was only causing a re-render when the cameras were changing, but there is probably more going on there than I thought. I had it setting a state variable there initially, which would then call setDefaultCamera outside the renderloop, when the state changed. But that resulted in a double render - once when the state updated and once when the camera updated. I'll have to look into this part more.